pytorch_lightning/utilities/deprecation.py

```diff
@@ -0,0 +1,73 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import inspect
+from functools import wraps
+from typing import Any, Callable, List, Tuple
+
+from pytorch_lightning.utilities import rank_zero_warn
+
+
+def get_func_arguments_and_types(func: Callable) -> List[Tuple[str, Tuple, Any]]:
+    """Parse function arguments, types and default values
+
+    Example:
+        >>> get_func_arguments_and_types(get_func_arguments_and_types)
+        [('func', typing.Callable, <class 'inspect._empty'>)]
+    """
+    func_default_params = inspect.signature(func).parameters
+    name_type_default = []
+    for arg in func_default_params:
+        arg_type = func_default_params[arg].annotation
+        arg_default = func_default_params[arg].default
+        name_type_default.append((arg, arg_type, arg_default))
+    return name_type_default
+
+
+def _deprecated(target: Callable, ver_deprecate: str = "", ver_remove: str = "") -> Callable:
```

Comment:
Just about the naming: Can we remove the underscore? Since this is also exposed to the user (even if they're not intended to use it directly

Reply:
I was following @carmocca suggestion from parallel PR
https://github.com/PyTorchLightning/pytorch-lightning/pull/6510#discussion_r594672430
but not strong preference here :]
cc: @PyTorchLightning/core-contributors

Reply:
I agree with @justusschock

Reply:
I'm indifferent here either, however if I had to choose I'd get rid of the underscore because as a decorator it doesn't look clean imo. Does make sense because you want it to be protected, but just looks ugly when used imo :)

Reply:
ok, lest go with _ now and eventually change later is it easier to go from protected to the public the other way around...

Reply:
ok, going with just `deprecated` 🐰